woocommerce/woocommerce-gutenberg-products-block
# Add to cart context provider

This PR implements the Add To Cart Context provider and hooks it up to the variation picker, quantity input, and add to cart button.

Closes #2856 - implemented updated context
Closes #2757 - made variation picker work with new context provider
Closes #2738 - implements events
Closes #2737 - notices are surfaced using store notices context

This is all heavily modelled upon the Checkout State Provider. It implements form state (IDLE, PRISTINE, DISABLED, PROCESSING) and exposes dispatch actions to set cart request data for the API to work. We can expand these APIs as we look further at extension needs and other product types, as well as the form-element registration system we've yet to implement, but that's the next step.

## AddToCartFormContextProvider

This provider wraps children with Add to Cart Form state context, validation context and includes a `FormSubmit` component. This is similar to the Checkout Processing Component and listens for events to submit the request to the API.

## Add To Cart Form Context

The context exposes data about the add to cart form, as well as current status of the form submission. Status includes:

- PRISTINE: Form is in it's initialized state.
- IDLE: When form state has changed but there is no activity happening.
- DISABLED: If the form cannot be submitted due to missing constraints, this status is assigned.
- BEFORE_PROCESSING: This is the state before form processing begins after the add to cart button has been pressed. Validation occurs at point point.
- PROCESSING: After BEFORE_PROCESSING status emitters have finished successfully.
- AFTER_PROCESSING: After server side processing is completed  this status is set.

Anything added to `requestParams` is sent along with the API request to add to cart; this is what the variation form is hooking into to set the variation ID and variation data.

```
/**
 * @typedef {Object} AddToCartFormContext
 *
 * @property {boolean}                        showFormElements   True if showing a full add to cart form.
 * @property {Object}                         product            The product object to add to the cart.
 * @property {number}                         quantity           Stores the quantity being added to the cart.
 * @property {number}                         minQuantity        Min quantity that can be added to the cart.
 * @property {number}                         maxQuantity        Max quantity than can be added to the cart.
 * @property {Object}                         requestParams      List of params to send to the API.
 * @property {boolean}                        isIdle             True when the form state has changed and has no activity.
 * @property {boolean}                        isDisabled         True when the form cannot be submitted.
 * @property {boolean}                        isProcessing       True when the form has been submitted and is being processed.
 * @property {boolean}                        isBeforeProcessing True during any observers executing logic before form processing (eg. validation).
 * @property {boolean}                        isAfterProcessing  True when form status is AFTER_PROCESSING.
 * @property {boolean}                        hasError           True when the form is in an error state. Whatever caused the error (validation/payment method) will likely have triggered a notice.
 * @property {AddToCartFormEventRegistration} eventRegistration  Event emitters that can be subscribed to.
 * @property {AddToCartFormDispatchActions}   dispatchActions    Various actions that can be dispatched for the add to cart form context data.
 */
```

Event emitters exist for the stages of form submission:

- `onAddToCartAfterProcessingWithSuccess` - Used to register a callback that will fire after form has been processed and there are no errors.
- `onAddToCartAfterProcessingWithError` - Used to register a callback that will fire when the form has been processed and has an error.
- `onAddToCartBeforeProcessing` - Used to register a callback that will fire when the form has been submitted before being sent off to the server.

Available dispatch actions include:

- `resetForm` - Dispatches an action that resets the form to a pristine state.
- `submitForm` - Dispatches an action that tells the form to submit.
- `setQuantity` - Dispatches an action that sets the quantity to the given value.
- `setRequestParams` - Dispatches an action that sets params for the add to cart request (key value pairs).
- `setHasError` - Dispatches an action that sets the status to having an error.
- `setAfterProcessing` - Dispatches an action that sets the status to after processing and also sets the response data accordingly.

**To test:**

- Create a single product block for a simple product. Check you can add various quantities to the cart without issue.
- Repeat the test with a variable product. The variation attribute picker should function and validate if missing attributes.
- Test the checkout flow - I moved some shared files (event emitters and validation)

I tested by adding all product types as single product blocks to a page with and without showFormElements enabled.
